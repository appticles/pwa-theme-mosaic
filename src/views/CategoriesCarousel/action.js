import config from 'react-global-configuration';

export const REQUEST_CATEGORIES = 'REQUEST_CATEGORIES';
export const RECEIVE_CATEGORIES = 'RECEIVE_CATEGORIES';

export const requestCategories = () => ({
  type: REQUEST_CATEGORIES,
});

export const receiveCategories = categories => ({
  type: RECEIVE_CATEGORIES,
  categories,
});

export const fetchCategories = (params = {}) => (dispatch) => {
  dispatch(requestCategories());

  let url;
  if (params && params.post) {
    url = config.get('export').categories + '?post=' + String(params.post);
  } else {
    url =
      config.get('export').categories +
      '?' +
      Object.keys(params)
        .map(k => k + '=' + encodeURIComponent(params[k]))
        .join('&');
  }
  return fetch(url)
    .then(response => response.json())
    .then(json => dispatch(receiveCategories(json)))
    .catch(() => {
      dispatch(receiveCategories([]));
    });
};
