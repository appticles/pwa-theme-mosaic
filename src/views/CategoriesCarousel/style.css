.carousel-container {
  background-color: #32394a;
  height: 100vh; }

.carousel-container .pusher,
.carousel-container .slick-slider,
.carousel-container .slick-list,
.carousel-container .slick-track,
.carousel-container .pusher div {
  height: 100%; }

.carousel-container .slick-side {
  height: 110%; }

.carousel-container .slick-dots {
  bottom: 9%;
  width: 90%;
  left: 5%;
  border-radius: 20px;
  opacity: 1; }

.carousel-container .slick-dots li {
  margin: 0 5px; }

.carousel-container .slick-dots li button::before {
  font-size: 15px;
  color: #fafafa; }

.carousel-container .slick-dots li.slick-active button::before {
  color: #fafafa; }

.carousel-container .pusher .ui.loader.active::before, .carousel-container .pusher .ui.loader.active::after {
  top: 50%; }

.ui.image.tiny.logo {
  position: absolute;
  z-index: 1000;
  top: 16px;
  left: 16px; }
