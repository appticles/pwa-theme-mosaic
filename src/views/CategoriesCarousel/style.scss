@import "../../utilities/variables";

.carousel-container {
  height: 100vh;
}

.carousel-container .pusher,
.carousel-container .slick-slider,
.carousel-container .slick-list,
.carousel-container .slick-track,
.carousel-container .pusher div {
  height: 100%;
}

.carousel-container .slick-side {
  height: 110%;
}

.carousel-container .slick-dots {
  bottom: 9%;
  width: 90%;
  left: 5%;
  border-radius: 20px;
  opacity: 1;
}

.carousel-container .slick-dots li {
  margin: 0 5px;
}
.carousel-container .slick-dots li button::before {
  font-size: 15px;
  color: $buttons-bg-color;
}
.carousel-container .slick-dots li.slick-active button::before {
  color: $buttons-bg-color;
}
