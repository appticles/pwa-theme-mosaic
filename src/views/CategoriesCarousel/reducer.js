import { combineReducers } from 'redux';
import _ from 'lodash';
import PropTypes from 'prop-types';
import { REQUEST_CATEGORIES, RECEIVE_CATEGORIES } from './action';

export const categoryPropType = PropTypes.shape({
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  slug: PropTypes.string.isRequired,
});

const items = (state = [], action) => {
  switch (action.type) {
  case REQUEST_CATEGORIES:
    return state;
  case RECEIVE_CATEGORIES:
    return _.unionBy(state, action.categories, 'id');
  default:
    return state;
  }
};
const isFetching = (state = 0, action) => {
  switch (action.type) {
  case REQUEST_CATEGORIES:
    return state + 1;
  case RECEIVE_CATEGORIES:
    return state - 1;
  default:
    return state;
  }
};

export const getCategories = state => state.items;
export const getCategoriesFetching = state => state.isFetching;
export default combineReducers({ items, isFetching });
