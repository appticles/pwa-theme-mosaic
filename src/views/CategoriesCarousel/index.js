import React, { Component } from 'react';
import Slider from 'react-slick';
import PropTypes from 'prop-types';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { withRouter } from 'react-router';
import { Sidebar, Loader, Image } from 'semantic-ui-react';
import config from 'react-global-configuration';

import { fetchCategories } from './action';
import { getCategories, getCategoriesFetching, categoryPropType } from './reducer';

import SideMenu from '../SideMenu/index';
import NavBar from '../../components/NavBar/index';
import { closeMenu } from '../../components/NavBar/action';

import CategoriesList from '../../components/CategoriesList/index';
import './style.css';

class CategoriesCarousel extends Component {
  constructor(props) {
    super(props);
    this.hideSidebar = this.hideSidebar.bind(this);
  }

  componentWillMount() {
    const { dispatch } = this.props;
    dispatch(fetchCategories());
  }
  createCategoriesList(homeChunkSize, regularChunkSize) {
    // get the subset of categories for home card
    const categoriesList = [];
    categoriesList.push(this.props.categories.slice(0, homeChunkSize));

    let i;
    for (i = homeChunkSize; i < this.props.categories.length; i += regularChunkSize) {
      categoriesList.push(this.props.categories.slice(i, i + regularChunkSize));
    }
    return categoriesList;
  }

  hideSidebar() {
    if (this.props.sideMenuVisible) {
      this.props.closeMenu();
    }
  }

  render() {
    const settings = {
      arrows: false,
      centerPadding: '50px',
      dots: true,
      infinite: false,
      speed: 500,
      slidesToShow: 1,
      slidesToScroll: 1,
    };

    const categoriesList = this.createCategoriesList(3, 5);

    return (
      <div className="carousel-container">
        <Sidebar.Pushable>
          <SideMenu />
          <Sidebar.Pusher dimmed={this.props.sideMenuVisible} onClick={this.hideSidebar}>
            <Image src={config.get('logo')} size="tiny" />
            <NavBar />
            {this.props.loading === 1 ? <Loader active /> : ''}
            <Slider {...settings}>
              {categoriesList.map((categoriesChunk, k) => (
                <div key={Math.random(k)} className="categories">
                  <CategoriesList categoriesChunk={categoriesChunk} />
                </div>
              ))}
            </Slider>
          </Sidebar.Pusher>
        </Sidebar.Pushable>
      </div>
    );
  }
}

CategoriesCarousel.propTypes = {
  dispatch: PropTypes.func.isRequired,
  loading: PropTypes.number.isRequired,
  categories: PropTypes.arrayOf(categoryPropType).isRequired,
  sideMenuVisible: PropTypes.bool.isRequired,
  closeMenu: PropTypes.func.isRequired,
};
const mapStateToProps = state => ({
  loading: getCategoriesFetching(state.categories),
  categories: getCategories(state.categories),
  sideMenuVisible: state.sideMenuVisible,
});
function mapDispatchToProps(dispatch) {
  return Object.assign({ dispatch }, bindActionCreators({ fetchCategories, closeMenu }, dispatch));
}
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(CategoriesCarousel));
