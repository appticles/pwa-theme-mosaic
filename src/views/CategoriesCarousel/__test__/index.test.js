import React from 'react';
import renderer from 'react-test-renderer';
import sinon from 'sinon';

import mockStore from '../../../__mocks__/redux-mock-store';
import CategoriesCarousel from '../index';

jest.mock('../components/CategoriesList/index', () => 'Categories list mock');
jest.mock('../../../components/NavBar/index', () => 'Navbar mock');
jest.mock('../../SideMenu/index', () => 'Page side menu mock');

jest.mock('react-slick', () => require.requireActual('../../../__mocks__/react-slick').default);
jest.mock('react-global-configuration', () => require.requireActual('../../../__mocks__/react-global-configuration').default);

describe('Categories carousel', () => {
  it('renders categories', () => {
    const store = mockStore({
      categories: {
        items: [
          {
            id: 32,
            name: 'Cat 1',
            slug: 'cat-1',
            link: 'http://localhost/cat-1',
          },
          {
            id: 43,
            name: 'Cat 2',
            slug: 'cat-2',
            link: 'http://localhost/cat-2',
          },
        ],
        isFetching: 0,
        loadMore: false,
      },
      sideMenuVisible: false,
    });
    expect(renderer.create(<CategoriesCarousel store={store} fetchCategories={jest.fn()} closeMenu={jest.fn()} />)).toMatchSnapshot();
  });

  it('renders loader', () => {
    const store = mockStore({
      categories: {
        items: [],
        isFetching: 1,
        loadMore: false,
      },
      sideMenuVisible: false,
    });
    expect(renderer.create(<CategoriesCarousel store={store} fetchCategories={jest.fn()} closeMenu={jest.fn()} />)).toMatchSnapshot();
  });

  it('calculates the number of pages and cards with no categories', () => {
    const componentInstanceNoCategories = renderer.create(
      <CategoriesCarousel.WrappedComponent
        dispatch={jest.fn()}
        loading={0}
        categories={[]}
        sideMenuVisible={false}
        closeMenu={jest.fn()}
        loadMore={false}
      />).getInstance();

    expect(componentInstanceNoCategories.getPageNumber()).toBe(0);
    expect(componentInstanceNoCategories.getNoCards()).toBe(0);
    expect(componentInstanceNoCategories.createCategoriesList(3, 5)).toEqual([[]]);
  });

  it('calculates the number of pages and cards for single card', () => {
    const categories = [
      { id: 1, name: 'Cat 1', slug: 'cat-1', link: 'http://localhost/cat-1' },
      { id: 2, name: 'Cat 2', slug: 'cat-2', link: 'http://localhost/cat-2' },
      { id: 3, name: 'Cat 3', slug: 'cat-3', link: 'http://localhost/cat-3' },
    ];

    const componentInstance = renderer.create(
      <CategoriesCarousel.WrappedComponent
        dispatch={jest.fn()}
        loading={0}
        categories={categories}
        sideMenuVisible={false}
        closeMenu={jest.fn()}
        loadMore={false}
      />).getInstance();

    expect(componentInstance.getPageNumber()).toBe(0);
    expect(componentInstance.getNoCards()).toBe(1);
    expect(componentInstance.createCategoriesList(3, 5)).toEqual([[...categories]]);
  });

  it('calculates the number of cards for two cards', () => {
    const moreCategories = [
      { id: 1, name: 'Cat 1', slug: 'cat-1', link: 'http://localhost/cat-1' },
      { id: 2, name: 'Cat 2', slug: 'cat-2', link: 'http://localhost/cat-2' },
      { id: 3, name: 'Cat 3', slug: 'cat-3', link: 'http://localhost/cat-3' },
      { id: 4, name: 'Cat 4', slug: 'cat-4', link: 'http://localhost/cat-4' },
      { id: 5, name: 'Cat 5', slug: 'cat-5', link: 'http://localhost/cat-5' },
      { id: 6, name: 'Cat 6', slug: 'cat-6', link: 'http://localhost/cat-6' },
      { id: 7, name: 'Cat 7', slug: 'cat-7', link: 'http://localhost/cat-7' },
    ];

    const componentInstance2 = renderer.create(
      <CategoriesCarousel.WrappedComponent
        dispatch={jest.fn()}
        loading={0}
        categories={moreCategories}
        sideMenuVisible={false}
        closeMenu={jest.fn()}
        loadMore={false}
      />).getInstance();

    expect(componentInstance2.getPageNumber()).toBe(0);
    expect(componentInstance2.getNoCards()).toBe(2);
    expect(componentInstance2.createCategoriesList(3, 5)).toEqual(
      [
        [moreCategories[0], moreCategories[1], moreCategories[2]],
        [moreCategories[3], moreCategories[4], moreCategories[5], moreCategories[6]],
      ],
    );
  });

  it('calculates the number of cards for multiple cards', () => {
    const moreCategories2 = [
      { id: 1, name: 'Cat 1', slug: 'cat-1', link: 'http://localhost/cat-1' },
      { id: 2, name: 'Cat 2', slug: 'cat-2', link: 'http://localhost/cat-2' },
      { id: 3, name: 'Cat 3', slug: 'cat-3', link: 'http://localhost/cat-3' },
      { id: 4, name: 'Cat 4', slug: 'cat-4', link: 'http://localhost/cat-4' },
      { id: 5, name: 'Cat 5', slug: 'cat-5', link: 'http://localhost/cat-5' },
      { id: 6, name: 'Cat 6', slug: 'cat-6', link: 'http://localhost/cat-6' },
      { id: 7, name: 'Cat 7', slug: 'cat-7', link: 'http://localhost/cat-7' },
      { id: 8, name: 'Cat 8', slug: 'cat-8', link: 'http://localhost/cat-8' },
      { id: 9, name: 'Cat 9', slug: 'cat-9', link: 'http://localhost/cat-9' },
      { id: 10, name: 'Cat 10', slug: 'cat-10', link: 'http://localhost/cat-10' },
      { id: 11, name: 'Cat 11', slug: 'cat-11', link: 'http://localhost/cat-11' },
      { id: 12, name: 'Cat 12', slug: 'cat-12', link: 'http://localhost/cat-12' },
      { id: 13, name: 'Cat 13', slug: 'cat-13', link: 'http://localhost/cat-13' },
      { id: 14, name: 'Cat 14', slug: 'cat-14', link: 'http://localhost/cat-14' },
    ];

    const componentInstance3 = renderer.create(
      <CategoriesCarousel.WrappedComponent
        dispatch={jest.fn()}
        loading={0}
        categories={moreCategories2}
        sideMenuVisible={false}
        closeMenu={jest.fn()}
        loadMore={false}
      />).getInstance();

    expect(componentInstance3.getPageNumber()).toBe(1);
    expect(componentInstance3.getNoCards()).toBe(4);
  });

  it('disable load more', () => {
    const dispatchSpy = sinon.spy(jest.fn());

    const moreCategories = [
      { id: 1, name: 'Cat 1', slug: 'cat-1', link: 'http://localhost/cat-1' },
      { id: 2, name: 'Cat 2', slug: 'cat-2', link: 'http://localhost/cat-2' },
      { id: 3, name: 'Cat 3', slug: 'cat-3', link: 'http://localhost/cat-3' },
    ];

    const componentInstance = renderer.create(
      <CategoriesCarousel.WrappedComponent
        dispatch={dispatchSpy}
        loading={0}
        categories={moreCategories}
        sideMenuVisible={false}
        closeMenu={jest.fn()}
        loadMore={false}
      />).getInstance();

    componentInstance.loadMore(1); // this call should not trigger dispatch

    // dispatch is also called once by the constructor
    expect(dispatchSpy.calledOnce).toBe(true);
  });

  it('load more categories', () => {
    const dispatchSpy = sinon.spy(jest.fn());

    const moreCategories = [
      { id: 1, name: 'Cat 1', slug: 'cat-1', link: 'http://localhost/cat-1' },
      { id: 2, name: 'Cat 2', slug: 'cat-2', link: 'http://localhost/cat-2' },
      { id: 3, name: 'Cat 3', slug: 'cat-3', link: 'http://localhost/cat-3' },
      { id: 4, name: 'Cat 4', slug: 'cat-4', link: 'http://localhost/cat-4' },
      { id: 5, name: 'Cat 5', slug: 'cat-5', link: 'http://localhost/cat-5' },
      { id: 6, name: 'Cat 6', slug: 'cat-6', link: 'http://localhost/cat-6' },
      { id: 7, name: 'Cat 7', slug: 'cat-7', link: 'http://localhost/cat-7' },
      { id: 8, name: 'Cat 8', slug: 'cat-8', link: 'http://localhost/cat-8' },
      { id: 9, name: 'Cat 9', slug: 'cat-9', link: 'http://localhost/cat-9' },
      { id: 10, name: 'Cat 10', slug: 'cat-10', link: 'http://localhost/cat-10' },
      { id: 11, name: 'Cat 11', slug: 'cat-11', link: 'http://localhost/cat-11' },
      { id: 12, name: 'Cat 12', slug: 'cat-12', link: 'http://localhost/cat-12' },
      { id: 13, name: 'Cat 13', slug: 'cat-13', link: 'http://localhost/cat-13' },
      { id: 14, name: 'Cat 14', slug: 'cat-14', link: 'http://localhost/cat-14' },
    ];

    const componentInstance = renderer.create(
      <CategoriesCarousel.WrappedComponent
        dispatch={dispatchSpy}
        loading={0}
        categories={moreCategories}
        sideMenuVisible={false}
        closeMenu={jest.fn()}
        loadMore
      />).getInstance();

    componentInstance.loadMore(2); // this call should not trigger dispatch
    componentInstance.loadMore(3);

    // dispatch is also called once by the constructor
    expect(dispatchSpy.calledTwice).toBe(true);
  });

  it('hides side bar', () => {
    const closeMenuSpy = sinon.spy(jest.fn());

    const componentInstance = renderer.create(
      <CategoriesCarousel.WrappedComponent
        dispatch={jest.fn()}
        loading={0}
        categories={[]}
        sideMenuVisible
        closeMenu={closeMenuSpy}
        loadMore={false}
      />).getInstance();

    componentInstance.hideSidebar();
    expect(closeMenuSpy.calledOnce).toBe(true);
  });
});
