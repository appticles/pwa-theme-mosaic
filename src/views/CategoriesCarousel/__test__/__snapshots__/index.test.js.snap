// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Categories carousel renders categories 1`] = `
<div
  className="carousel-container"
>
  <div
    className="pushable"
  >
    <Page side menu mock />
    <div
      className="pusher"
      onClick={[Function]}
    >
      <img
        alt={undefined}
        className="ui image logo"
        height={undefined}
        src="logo"
        width={undefined}
      />
      <Navbar mock />
    </div>
  </div>
</div>
`;

exports[`Categories carousel renders loader 1`] = `
<div
  className="carousel-container"
>
  <div
    className="pushable"
  >
    <Page side menu mock />
    <div
      className="pusher"
      onClick={[Function]}
    >
      <img
        alt={undefined}
        className="ui image logo"
        height={undefined}
        src="logo"
        width={undefined}
      />
      <Navbar mock />
      <div
        className="ui active loader"
      />
    </div>
  </div>
</div>
`;
